ql/src/java/org/apache/hadoop/hive/ql/parse/repl/metric/MetricCollector.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.parse.repl.metric;
+
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.parse.repl.metric.event.ReplicationMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * MetricCollector.
+ * In memory collection of metrics
+ */
+public class MetricCollector {
+  private Logger LOG = LoggerFactory.getLogger(MetricCollector.class);
+  private Map<Long, ReplicationMetric> metricMap = new ConcurrentHashMap<>();
+  private final long maxSize;
+
+  private static volatile MetricCollector instance;
```

Comment:
volatile is not needed here.

Reply:
Why? You should use volatile on the double-checked locking idiom when creating mutable shared singletons.

Reply:
The synchronized block is the only thing required. It will ensure both atomicity and visibility.
Details here:https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4.5

Reply:
https://stackoverflow.com/questions/11639746/what-is-the-point-of-making-the-singleton-instance-volatile-while-using-double-l

Reply:
I would go with official java language specification rather as this topic is often confusing. jls says:
If an action x synchronizes-with a following action y, then we also have hb(x, y). So you don't have visibility issue here. And hence volatile will not be needed. You can keep it if still not convinced :-)

Reply:
Ok, sure. let's keep it. Actually the problem here is different than the visibility one. My mind was occupied with that visibility issue normally we see in primitive variable in multithreaded settings. Here the issue is with possible partial state of the object, which also might not have happened if all the members were declared final. Actually, the map we are using inside can be made final but until then it has issue and volatile will solve it.

Reply:
That is the reason using volatile in singleton pattern is a standard practise